distribution/tools/launchers/src/main/java/org/elasticsearch/tools/launchers/MachineDependentHeap.java

```diff
@@ -0,0 +1,247 @@
+        MachineNodeRole nodeRole = NodeRoleParser.parse(config);
+
+        try {
+            long availableSystemMemory = systemMemoryInfo.availableSystemMemory();
+            return options(nodeRole.heap(availableSystemMemory));
+        } catch (SystemMemoryInfo.SystemMemoryInfoException e) {
+            // If unable to determine system memory (ex: incompatible jdk version) fallback to defaults
+            return options(DEFAULT_HEAP_SIZE_MB);
+        }
+    }
+
+    private static List<String> options(int heapSize) {
+        return List.of("-Xms" + heapSize + "m", "-Xmx" + heapSize + "m");
+    }
+
+    /**
+     * Parses role information from elasticsearch.yml and determines machine node role.
+     */
+    static class NodeRoleParser {
+        private static final Set<String> LEGACY_ROLE_SETTINGS = Set.of(
+            "node.master",
+            "node.ingest",
+            "node.data",
+            "node.voting_only",
+            "node.ml",
+            "node.transform",
+            "node.remote_cluster_client"
+        );
+
+        @SuppressWarnings("unchecked")
+        public static MachineNodeRole parse(InputStream config) {
+            Yaml yaml = new Yaml();
+            Map<String, Object> root;
+            try {
+                root = yaml.load(config);
+            } catch (ClassCastException ex) {
+                // Strangely formatted config, so just return defaults and let startup settings validation catch the problem
+                return MachineNodeRole.UNKNOWN;
+            } catch (YAMLException ex) {
+                throw new IllegalStateException("Unable to parse elasticsearch.yml:", ex);
+            }
+
+            if (root != null) {
+                Map<String, Object> map = flatten(root, null);
+
+                if (hasLegacySettings(map.keySet())) {
+                    // We don't attempt to auto-determine heap if legacy role settings are used
+                    return MachineNodeRole.UNKNOWN;
+                } else {
+                    List<String> roles = null;
+                    try {
+                        if (map.containsKey("node.roles")) {
+                            roles = (List<String>) map.get("node.roles");
+                        }
+                    } catch (ClassCastException ex) {
+                        throw new IllegalStateException("Unable to parse elasticsearch.yml. Expected 'node.roles' to be a list.");
+                    }
+
+                    if (roles == null || roles.isEmpty()) {
+                        // If roles are missing or empty (coordinating node) assume defaults and consider this a data node
+                        return MachineNodeRole.DATA;
+                    } else if (containsOnly(roles, "master", "voting_only")) {
+                        return MachineNodeRole.MASTER_ONLY;
+                    } else if (containsOnly(roles, "ml", "voting_only")) {
+                        return MachineNodeRole.ML_ONLY;
```

Comment:
Why is voting only included here? And what about transform nodes? Can we check with the ML team how those should be handled?

Reply:
Not sure what was intended here, but I've removed the voting_only from both master and ML cases.

Reply:
After looking through the unit tests, it appears this was intended to support an ML node being a voting only node. However, I think there was confusion in what voting_only does, since it actually requires the master role as well. So I think removing them was correct.

Reply:
The intention here is that any node can have the voting_only role as well, and I believe cloud will do this to ensure there is an available tiebreaker node. Essentially, for the purposes of assigning memory based on the amount of _work_ the node will do, the `voting_only` node is irrelevant, and if an ML node has the role, it should still be considered ML_ONLY. The scenario with master is much less likely however so we can probably remove it.

Reply:
`voting_only` must be used in combination with `master`. So it's not possible to have ml+voting_only and nothing else, it would be ml+master+voting_only. For a master+voting_only, you would never have this, since there would be no point to not make the node a normal master eligible node. The point of voting only is to use in combination with other node roles yet do a little extra work to tiebreak.

Reply:
Ah, understood.